velox/exec/tests/MultiFragmentTest.cpp

```diff
@@ -787,3 +787,133 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
+      DriverCtx* driverCtx,
+      std::shared_ptr<const CustomSlowNode> slowNode)
+      : Operator(
+            driverCtx,
+            slowNode->outputType(),
+            operatorId,
+            slowNode->id(),
+            "CustomSlowOperator") {}
+
+  void addInput(RowVectorPtr input) override {
+    inputs_.emplace_back(std::move(input));
+  }
+
+  bool needsInput() const override {
+    return inputs_.empty();
+  }
+
+  RowVectorPtr getOutput() override {
+    if (inputs_.empty() || consumedOneBatch) {
+      return nullptr;
+    }
+    auto result = inputs_.back();
+    inputs_.pop_back();
+    consumedOneBatch = true;
+    return result;
+  }
+
+  void noMoreInput() override {
+    Operator::noMoreInput();
+  }
+
+  BlockingReason isBlocked(ContinueFuture* future) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() override {
+    return false;
+  }
+
+ private:
+  std::vector<RowVectorPtr> inputs_;
+  bool consumedOneBatch{false};
+};
+
+class CustomSlowOperatorTranslator : public Operator::PlanNodeTranslator {
+  std::unique_ptr<Operator>
+  toOperator(DriverCtx* ctx, int32_t id, const core::PlanNodePtr& node) {
+    if (auto slowNode = std::dynamic_pointer_cast<const CustomSlowNode>(node)) {
+      return std::make_unique<CustomSlowOperator>(id, ctx, slowNode);
+    }
+    return nullptr;
+  }
+};
+
+TEST_F(MultiFragmentTest, exchangeDestruction) {
+  // This unit test tests the proper destruction of ExchangeClient upon
+  // task destruction.
+  Operator::registerOperator(std::make_unique<CustomSlowOperatorTranslator>());
+
+  // Set small size so that ExchangeSource will conduct multiple rounds of
+  // pulls of data while CustomSlowOperator will only consume one batch (data
+  // from a single pull of ExchangeSource). This makes sure we always have
+  // data buffered at ExchangeQueue in Exchange operator, which this test
+  // requires.
+  configSettings_[core::QueryConfig::kMaxPartitionedOutputBufferSize] = "2048";
+  setupSources(10, 1000);
+  auto leafTaskId = makeTaskId("leaf", 0);
+  core::PlanNodePtr leafPlan;
+
+  leafPlan = PlanBuilder()
+                 .tableScan(rowType_)
+                 .project({"c0 % 10 AS c0", "c1"})
+                 .partitionedOutput({}, 1)
+                 .planNode();
+
+  auto leafTask = makeTask(leafTaskId, leafPlan, 0);
+  Task::start(leafTask, 1);
+  addHiveSplits(leafTask, filePaths_);
+
+  core::PlanNodePtr rootPlan;
+  rootPlan = PlanBuilder()
+                 .exchange(leafPlan->outputType())
+                 .addNode([&leafPlan](std::string id, core::PlanNodePtr node) {
+                   return std::make_shared<CustomSlowNode>(id, std::move(node));
+                 })
+                 .partitionedOutput({}, 1)
+                 .planNode();
+
+  auto rootTask = makeTask("root-task", rootPlan, 0);
+  Task::start(rootTask, 1);
+  addRemoteSplits(rootTask, {leafTaskId});
+
+  waitForTaskCompletion(rootTask.get(), 1'000'000);
```

Comment:
What do we expect to happen? I assume task will not complete. Should we assert it hasn't completed?